### DISCUSSION TITLE
Questions about the LocalAI feature in Nextcloud

### DISCUSSION THREAD

**Discussion by gbertolottiKS** [nextcloud/all-in-one#4552] [Questions] [answered]
Hi.
We have some questions about the AI feature in Nextcloud, especially regarding LocalAI:

1. Is there a guide on how to evaluate the models available on LocalAI (apart from downloading everyone and trying them)? Which one is better, e.g. for writing code, and which is better for understanding documents' content? More simply, which is usable for text generation, and which is for image generation? How do we distinguish them?

2. Are all the models available here (https://github.com/go-skynet/model-gallery/blob/main/index.yaml)? Is every other model not present on that page unusable?

3. We have a VPS with 24GB of RAM and 12 vCore, but we noticed that the result came very slowly (around 5 minutes after the prompt): is it normal? Are there some steps we can take to improve these timings? We know the real difference comes from the GPU, but unfortunately we don't have one.

4. We are not able to generate images: every time, it fails. We are using the url suggested by the README (github:go-skynet/model-gallery/stablediffusion.yaml). Below the raw log:

```{"reqId":"DRoppBXRkmIaXp5ZtV51","level":2,"time":"2024-04-15T15:36:15+00:00","remoteAddr":"XXX","user":"admin","app":"integration_openai","method":"POST","url":"/ocs/v2.php/apps/assistant/api/v1/i/process_prompt","message":"OpenAI/LocalAI's text to image generation failed with: API request error: rpc error: code = Unavailable desc = error reading from server: EOF","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:124.0) Gecko/20100101 Firefox/124.0","version":"28.0.4.1","exception":{"Exception":"Exception","Message":"API request error: rpc error: code = Unavailable desc = error reading from server: EOF","Code":500,"Trace":[{"file":"/var/www/html/custom_apps/integration_openai/lib/Service/OpenAiAPIService.php","line":519,"function":"request","class":"OCA\\OpenAi\\Service\\OpenAiAPIService","type":"->","args":["admin","images/generations",["a frog with a cap","1024x1024"],"POST"]},{"file":"/var/www/html/custom_apps/integration_openai/lib/TextToImage/TextToImageProvider.php","line":47,"function":"requestImageCreation","class":"OCA\\OpenAi\\Service\\OpenAiAPIService","type":"->","args":["admin","a frog with a cap",1]},{"file":"/var/www/html/lib/private/TextToImage/Manager.php","line":161,"function":"generate","class":"OCA\\OpenAi\\TextToImage\\TextToImageProvider","type":"->","args":["a frog with a cap",[null]]},{"file":"/var/www/html/lib/private/TextToImage/Manager.php","line":237,"function":"runTask","class":"OC\\TextToImage\\Manager","type":"->","args":[["OCP\\TextToImage\\Task"]]},{"file":"/var/www/html/custom_apps/assistant/lib/Service/Text2Image/Text2ImageHelperService.php","line":104,"function":"runOrScheduleTask","class":"OC\\TextToImage\\Manager","type":"->","args":[["OCP\\TextToImage\\Task"]]},{"file":"/var/www/html/custom_apps/assistant/lib/Controller/Text2ImageApiController.php","line":55,"function":"processPrompt","class":"OCA\\TpAssistant\\Service\\Text2Image\\Text2ImageHelperService","type":"->","args":["assistant","","a frog with a cap",1,false,"admin",true,false]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":230,"function":"processPrompt","class":"OCA\\TpAssistant\\Controller\\Text2ImageApiController","type":"->","args":["assistant","","a frog with a cap",1,false,true,false]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":137,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[["OCA\\TpAssistant\\Controller\\Text2ImageApiController"],"processPrompt"]},{"file":"/var/www/html/lib/private/AppFramework/App.php","line":184,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[["OCA\\TpAssistant\\Controller\\Text2ImageApiController"],"processPrompt"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":315,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\TpAssistant\\Controller\\Text2ImageApiController","processPrompt",["OC\\AppFramework\\DependencyInjection\\DIContainer"],["v1","ocs.assistant.Text2ImageApi.processPrompt"]]},{"file":"/var/www/html/ocs/v1.php","line":65,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/ocsapp/apps/assistant/api/v1/i/process_prompt"]},{"file":"/var/www/html/ocs/v2.php","line":23,"args":["/var/www/html/ocs/v1.php"],"function":"require_once"}],"File":"/var/www/html/custom_apps/integration_openai/lib/Service/OpenAiAPIService.php","Line":790,"message":"OpenAI/LocalAI's text to image generation failed with: API request error: rpc error: code = Unavailable desc = error reading from server: EOF","exception":[],"CustomMessage":"OpenAI/LocalAI's text to image generation failed with: API request error: rpc error: code = Unavailable desc = error reading from server: EOF"},"id":"661d497b7bc77"}```

5. We activated the Docker proxy container and downloaded the Context Chat External app, but we couldn't make use of the Context Chat feature. Logs below:

   ```{"reqId":"BddcFHfgu8zXnPDavIwk","level":3,"time":"2024-04-15T15:39:30+00:00","remoteAddr":"","user":"--","app":"app_api","method":"","url":"--","message":"Error during request to ExApp context_chat_backend: Server error: `PUT http://context_chat_backend:23000/loadSources` resulted in a `500 Internal Server Error` response:\nTraceback (most recent call last):\n File \"/usr/local/lib/python3.11/dist-packages/starlette/middleware/errors.py\", line (truncated...)\n","userAgent":"--","version":"28.0.4.1","exception":{"Exception":"GuzzleHttp\\Exception\\ServerException","Message":"Server error: `PUT http://context_chat_backend:23000/loadSources` resulted in a `500 Internal Server Error` response:\nTraceback (most recent call last):\n File \"/usr/local/lib/python3.11/dist-packages/starlette/middleware/errors.py\", line (truncated...)\n","Code":500,"Trace":[{"file":"/var/www/html/3rdparty/guzzlehttp/guzzle/src/Middleware.php","line":69,"function":"create","class":"GuzzleHttp\\Exception\\RequestException","type":"::","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/3rdparty/guzzlehttp/promises/src/Promise.php","line":204,"function":"GuzzleHttp\\{closure}","class":"GuzzleHttp\\Middleware","type":"::","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/3rdparty/guzzlehttp/promises/src/Promise.php","line":153,"function":"callHandler","class":"GuzzleHttp\\Promise\\Promise","type":"::","args":[1,"*** sensitive parameters replaced ***","*** sensitive parameters replaced ***"]},{"file":"/var/www/html/3rdparty/guzzlehttp/promises/src/TaskQueue.php","line":48,"function":"GuzzleHttp\\Promise\\{closure}","class":"GuzzleHttp\\Promise\\Promise","type":"::","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/3rdparty/guzzlehttp/promises/src/Promise.php","line":248,"function":"run","class":"GuzzleHttp\\Promise\\TaskQueue","type":"->","args":[true]},{"file":"/var/www/html/3rdparty/guzzlehttp/promises/src/Promise.php","line":224,"function":"invokeWaitFn","class":"GuzzleHttp\\Promise\\Promise","type":"->","args":[]},{"file":"/var/www/html/3rdparty/guzzlehttp/promises/src/Promise.php","line":269,"function":"waitIfPending","class":"GuzzleHttp\\Promise\\Promise","type":"->","args":[]},{"file":"/var/www/html/3rdparty/guzzlehttp/promises/src/Promise.php","line":226,"function":"invokeWaitList","class":"GuzzleHttp\\Promise\\Promise","type":"->","args":[]},{"file":"/var/www/html/3rdparty/guzzlehttp/promises/src/Promise.php","line":62,"function":"waitIfPending","class":"GuzzleHttp\\Promise\\Promise","type":"->","args":[]},{"file":"/var/www/html/3rdparty/guzzlehttp/guzzle/src/Client.php","line":187,"function":"wait","class":"GuzzleHttp\\Promise\\Promise","type":"->","args":[]},{"file":"/var/www/html/lib/private/Http/Client/Client.php","line":337,"function":"request","class":"GuzzleHttp\\Client","type":"->","args":["put","http://context_chat_backend:23000/loadSources",["/mnt/ncdata/files_external/rootcerts.crt","7200",[["sources","files__default: 10",null,["admin","/admin/files/Photos/Readme.md","text/markdown",1712844131,"files__default"]]],["2.4.0","context_chat_backend","2.0.1","OmtrQjltRGtKdm5XcWJ3NHdRempmV0o4OU9JSHE0bmV2MUV2Y0RCZE5ucTMvU0tNZTIxaFB5dlg5dnJ3VVV4ZGxSL3pUVUVXSmtoZjVHRDhSd2lCMGhCM1pZTEpRSUJIblVTYVFYVzM2NHljdHc3NVE1bUdXVzYxY3FpOFdDQ1hn","CLI","And 3 more entries, set log level to debug to see all entries"],[true],"And 1 more entries, set log level to debug to see all entries"]]},{"file":"/var/www/html/custom_apps/app_api/lib/Service/AppAPIService.php","line":108,"function":"put","class":"OC\\Http\\Client\\Client","type":"->","args":["http://context_chat_backend:23000/loadSources",["7200",[["sources","files__default: 10",null,["admin","/admin/files/Photos/Readme.md","text/markdown",1712844131,"files__default"]]],["2.4.0","context_chat_backend","2.0.1","OmtrQjltRGtKdm5XcWJ3NHdRempmV0o4OU9JSHE0bmV2MUV2Y0RCZE5ucTMvU0tNZTIxaFB5dlg5dnJ3VVV4ZGxSL3pUVUVXSmtoZjVHRDhSd2lCMGhCM1pZTEpRSUJIblVTYVFYVzM2NHljdHc3NVE1bUdXVzYxY3FpOFdDQ1hn","CLI","And 1 more entries, set log level to debug to see all entries"],[true]]]},{"file":"/var/www/html/custom_apps/app_api/lib/Service/AppAPIService.php","line":89,"function":"requestToExAppInternal","class":"OCA\\AppAPI\\Service\\AppAPIService","type":"->","args":[["OCA\\AppAPI\\Db\\ExApp",1],"PUT","http://context_chat_backend:23000/loadSources",["SensitiveParameterValue"]]},{"file":"/var/www/html/custom_apps/app_api/lib/PublicFunctions.php","line":36,"function":"requestToExApp","class":"OCA\\AppAPI\\Service\\AppAPIService","type":"->","args":[["OCA\\AppAPI\\Db\\ExApp",1],"/loadSources","*** sensitive parameters replaced ***","PUT",[["sources","files__default: 10",null,["admin","/admin/files/Photos/Readme.md","text/markdown",1712844131,"files__default"]]],["7200",[["sources","files__default: 10",null,["admin","/admin/files/Photos/Readme.md","text/markdown",1712844131,"files__default"]]]],"**_ sensitive parameters replaced _**"]},{"file":"/var/www/html/custom_apps/context_chat/lib/Service/LangRopeService.php","line":100,"function":"exAppRequest","class":"OCA\\AppAPI\\PublicFunctions","type":"->","args":["context_chat_backend","/loadSources","**_ sensitive parameters replaced _**","PUT",[["sources","files__default: 10",null,["admin","/admin/files/Photos/Readme.md","text/markdown",1712844131,"files__default"]]],["7200",[["sources","files__default: 10",null,["admin","/admin/files/Photos/Readme.md","text/markdown",1712844131,"files__default"]]]]]},{"file":"/var/www/html/custom_apps/context_chat/lib/Service/LangRopeService.php","line":199,"function":"requestToExApp","class":"OCA\\ContextChat\\Service\\LangRopeService","type":"->","args":["/loadSources","PUT",[["sources","files__default: 10",null,["admin","/admin/files/Photos/Readme.md","text/markdown",1712844131,"files__default"]]],"multipart/form-data"]},{"file":"/var/www/html/custom_apps/context_chat/lib/BackgroundJobs/IndexerJob.php","line":146,"function":"indexSources","class":"OCA\\ContextChat\\Service\\LangRopeService","type":"->","args":[[["OCA\\ContextChat\\Type\\Source","admin","files__default: 10","/admin/files/Photos/Readme.md",null,1712844131,"text/markdown","files__default"]]]},{"file":"/var/www/html/custom_apps/context_chat/lib/BackgroundJobs/IndexerJob.php","line":79,"function":"index","class":"OCA\\ContextChat\\BackgroundJobs\\IndexerJob","type":"->","args":[[["OCA\\ContextChat\\Db\\QueueFile",1],["OCA\\ContextChat\\Db\\QueueFile",2],["OCA\\ContextChat\\Db\\QueueFile",3],["OCA\\ContextChat\\Db\\QueueFile",4],["OCA\\ContextChat\\Db\\QueueFile",5],"And 20 more entries, set log level to debug to see all entries"]]},{"file":"/var/www/html/lib/public/BackgroundJob/Job.php","line":81,"function":"run","class":"OCA\\ContextChat\\BackgroundJobs\\IndexerJob","type":"->","args":[[1,1]]},{"file":"/var/www/html/lib/public/BackgroundJob/TimedJob.php","line":102,"function":"start","class":"OCP\\BackgroundJob\\Job","type":"->","args":[["OC\\BackgroundJob\\JobList"]]},{"file":"/var/www/html/lib/public/BackgroundJob/TimedJob.php","line":92,"function":"start","class":"OCP\\BackgroundJob\\TimedJob","type":"->","args":[["OC\\BackgroundJob\\JobList"]]},{"file":"/var/www/html/cron.php","line":152,"function":"execute","class":"OCP\\BackgroundJob\\TimedJob","type":"->","args":[["OC\\BackgroundJob\\JobList"],["OC\\Log"]]}],"File":"/var/www/html/3rdparty/guzzlehttp/guzzle/src/Exception/RequestException.php","Line":113,"message":"Error during request to ExApp context_chat_backend: Server error: `PUT http://context_chat_backend:23000/loadSources` resulted in a `500 Internal Server Error` response:\nTraceback (most recent call last):\n File \"/usr/local/lib/python3.11/dist-packages/starlette/middleware/errors.py\", line (truncated...)\n","exception":[],"CustomMessage":"Error during request to ExApp context_chat_backend: Server error: `PUT http://context_chat_backend:23000/loadSources` resulted in a `500 Internal Server Error` response:\nTraceback (most recent call last):\n File \"/usr/local/lib/python3.11/dist-packages/starlette/middleware/errors.py\", line (truncated...)\n"},"id":"661d4a336207f"}```

```{"reqId":"BddcFHfgu8zXnPDavIwk","level":3,"time":"2024-04-15T15:39:31+00:00","remoteAddr":"","user":"--","app":"context_chat","method":"","url":"--","message":"Error during request to ExApp (context_chat_backend): Server error: `POST http://context_chat_backend:23000/query` resulted in a `500 Internal Server Error` response:\n{\"error\":\"Error: LLM not initialised\"}\n","userAgent":"--","version":"28.0.4.1","data":{"app":"context_chat"},"id":"661d4a3361d94"}```

Thanks in advance

**Comment by szaimen** [comment, marked as answer]
> 1. Is there a guide on how to evaluate the models available on LocalAI (apart from downloading everyone and trying them)? Which one is better, e.g. for writing code, and which is better for understanding documents' content? More simply, which is usable for text generation, and which is for image generation? How do we distinguish them?

I cannot tell unfortunately. You probably need to find out yourself.

> 2. Are all the models available here (https://github.com/go-skynet/model-gallery/blob/main/index.yaml)? Is every other model not present on that page unusable?

Yeah, in theory should all be usable but note that we ship the base container which does not have all backends installed by default so that will probably limit the amount of models that can be used.

> 3. We have a VPS with 24GB of RAM and 12 vCore, but we noticed that the result came very slowly (around 5 minutes after the prompt): is it normal? Are there some steps we can take to improve these timings? We know the real difference comes from the GPU, but unfortunately we don't have one.

Yes, this sounds normal and is one of the huge problems of any AI in the moment: it just needs a lot of resources.

> 4. We are not able to generate images: every time, it fails. We are using the url suggested by the README (github:go-skynet/model-gallery/stablediffusion.yaml). Below the raw log:

Can you check if this is already reported here? https://github.com/szaimen/aio-local-ai/issues

> 5\. We activated the Docker proxy container and downloaded the Context Chat External app, but we couldn't make use of the Context Chat feature. Logs below:

This should probably reported here: https://github.com/nextcloud/context_chat_backend
